### PR TITLE
use a spatial index in EmpiricalVariogram

### PR DESCRIPTION
**Pull request by visr**
If there are many points involved, and the search radius is small relative to the complete domain, using a spatial index is more efficient than iterating over half the points. For one real world dataset this sped up the algorithm about 20x, though of course it greatly varies. Results are identical.

Besides the spatial index, adding an early exit for if `h > hmax` already speeds up the code, avoiding the expensive `v` calculation.

I haven't yet added tests, since we probably want to discuss this first.

There are still some other optimisations that I didn't include here, such as `z₁ = sdata[var₁]; z₂ = sdata[var₂]` creating two distinct vectors even if `var₁ == var₂`, and taking `z₁[j]` and `z₂[j]` `getindex` out of the inner loop.

### CODE REVIEW

**Comment by juliohm** [comment]
Nice! I love to see performance improvements 💯 Let's evolve the PR to make it faster and more widely tested. Will add some comments to the current version.

File: src/empirical.jl [hunk 1]
```diff
@@ -51,18 +51,43 @@ struct EmpiricalVariogram
     sums   = zeros(nlags)
     counts = zeros(Int, nlags)
 
+    # decide whether to construct a spatial index
+    # coordinates need to be AbstractFloat for NearestNeighbors to work
+    # https://github.com/KristofferC/NearestNeighbors.jl/issues/13
+    isfloat = eltype(coordinates(sdata)) <: AbstractFloat
```
Comment (juliohm) [inline src/empirical.jl]: You can use `coordtype(sdata)` to get the type of the coordinates without building the array.

File: src/empirical.jl [hunk 2]
```diff
@@ -51,18 +51,43 @@ struct EmpiricalVariogram
...
     xi = MVector{ndims(sdata),coordtype(sdata)}(undef)
     xj = MVector{ndims(sdata),coordtype(sdata)}(undef)
 
-    @inbounds (
-    for j in 1:npts
+    @inbounds for j in 1:npts
       coordinates!(xj, sdata, j)
-      for i in j+1:npts
+
+      if use_spatial_index
```
Comment (juliohm) [inline src/empirical.jl]: We could move this check that is happening inside the loop to outside. Actually we should refactor the code to look like two major branches: one branch in which we use the nearest neighbour search and another branch that doesn't. Could you refactor the code so that it looks more compact? Something like:

```julia
if is float && isbig
  # construct the variogram with neighbour search
else
  # construct the variogram with naive loops
end
```

File: src/empirical.jl [hunk 3]
```diff
@@ -51,18 +51,43 @@ struct EmpiricalVariogram
...
+        pt_indices = search(xj, neighsearcher)
+      else
+        pt_indices = j+1:npts
+      end
+
+      for i in pt_indices
+        i <= j && continue  # spatial index include these
```
Comment (juliohm) [inline src/empirical.jl]: This check for example should only exist in the branch that computes the variogram with neighbourhood search.

File: src/empirical.jl [hunk 4]
```diff
@@ -51,18 +51,43 @@ struct EmpiricalVariogram
...
         coordinates!(xi, sdata, i)
 
         # evaluate lag and (cross-)variance
         h = evaluate(distance, xi, xj)
+        # early exit if out of range
+        h > hmax && continue
```
Comment (juliohm) [inline src/empirical.jl]: That is a nice catch. We should benchmark it to see if the check is faster than computing the floating point `v` regardless.

**Comment by visr** [comment]
Nice, thanks for the reply and comments. I'll work on updates in a few days.

File: src/empirical.jl (hunk 1 above)
Comment (visr) [reply src/empirical.jl]: Good tip, done in 0009ea940c63a829c0238850f89467d7e5bf8557

File: src/empirical.jl (hunk 2 above)
Comment (visr) [reply src/empirical.jl]: Yes, I refactored it to this structure in 0009ea940c63a829c0238850f89467d7e5bf8557. It does result in some more duplicated lines, if you have further ideas on how to structure it, feel free to also push to the branch.

Note that this check, even though it was inside the loop, was not expensive. I'm not sure, but it might have been optimised out completely.

File: src/empirical.jl (hunk 3 above)
Comment (visr) [reply src/empirical.jl]: You're right, I did this in 0009ea940c63a829c0238850f89467d7e5bf8557. Don't think it's an expensive check, but indeed not needed.

File: src/empirical.jl (hunk 4 above)
Comment (visr) [reply src/empirical.jl]: Interestingly the `v` calculation is the most expensive, it spends about 50% of its time there. So avoiding it pays off, especially if it is true often. Forcing it manually to take the no spatial index branch, this:

```julia
using BenchmarkTools
using GeoStats

n = 10_000
X = rand(2, n)
z = rand(n)

@btime EmpiricalVariogram($X, $z, nlags=20, maxlag=0.05)
```

Will show:
```
no spatial index with hmax early exit
1.002 s (4840077 allocations: 74.02 MiB)
no spatial index without hmax early exit
10.080 s (540516782 allocations: 8.05 GiB)
```

Comment (juliohm) [reply src/empirical.jl]: Nice to see these speedups. 💯

File: src/empirical.jl [hunk 5]
```diff
@@ -55,26 +55,65 @@ struct EmpiricalVariogram
...
+    # decide whether to construct a spatial index
+    # coordinates need to be AbstractFloat for NearestNeighbors to work
+    # https://github.com/KristofferC/NearestNeighbors.jl/issues/13
+    isfloat = coordtype(sdata) <: AbstractFloat
+    # quite arbitrary, but for small datasets setting up the index is too much overhead
+    # it really also depends on what fraction of points are within hmax
+    isbig = npts > 1000
+
+    if isfloat && isbig
+      # construct the variogram with neighbour search
```
Comment (juliohm) [inline src/empirical.jl]: What about creating an additional function called `spatial_index_variogram` for this branch, and another function `simple_loop_variogram`? These could then be more easily tested in the test suite.

File: src/empirical.jl [hunk 6]
```diff
@@ -55,26 +55,65 @@ struct EmpiricalVariogram
...
+          # bin (or lag) where to accumulate result
+          lag = ceil(Int, h / Δh)
+
+          if lag ≤ nlags && !ismissing(v) && !isnan(v)
+            sums[lag] += v
+            counts[lag] += 1
+          end
+        end
+      end
+
```
Comment (juliohm) [inline src/empirical.jl]: Extra empty line

File: test/runtests.jl [hunk 7]
```diff
@@ -14,11 +14,7 @@ ENV["GKSwstype"] = "100"
 islinux = Sys.islinux()
 istravis = "TRAVIS" ∈ keys(ENV)
 datadir = joinpath(@__DIR__,"data")
-visualtests = !istravis || (istravis && islinux)
-if !istravis
-  Pkg.add("Gtk")
-  using Gtk
-end
+visualtests = false
```
Comment (juliohm) [inline test/runtests.jl]: Please undo this change in the final iteration so that we can merge :)

Comment (visr) [reply test/runtests.jl]: Woops, snuck in. Removed.

**Comment by visr** [comment]
Ok, made the rest of the changes and added tests, this should be good to go from my side.

I see that a test is failing, is this unrelated noise?

```
Warning: Image did not match reference image /home/travis/build/juliohm/Variography.jl/test/data/Fitting.png. err: Images differ.  Difference: 0.022093271946049414  tolerance: 0.02
```

File: src/empirical.jl [hunk 8]
```diff
@@ -2,6 +2,57 @@
 # Licensed under the ISC License. See LICENSE in the project root.
 # ------------------------------------------------------------------
 
+"Internal function used to calculate the empirical variogram using a spatial index"
+function spatial_index_variogram!(sums, counts, sdata, xi, xj, distance, nlags, Δh, hmax, z₁, z₂, npts)
```
Comment (juliohm) [inline src/empirical.jl]: Can we shorten the list of arguments a bit? Even if we repeat some code inside each function, this list of argument seems too huge.

File: src/empirical.jl [hunk 9]
```diff
@@ -30,7 +82,7 @@ struct EmpiricalVariogram
   ordinate::Vector{Float64}
   counts::Vector{Int}
 
-  function EmpiricalVariogram(sdata, var₁, var₂, nlags, maxlag, distance)
+  function EmpiricalVariogram(sdata, var₁, var₂, nlags, maxlag, distance, nearestsearch)
```
Comment (juliohm) [inline src/empirical.jl]: I think we should make this a keyword argument as opposed to a positional argument.

**Comment by juliohm** [comment]
> Ok, made the rest of the changes and added tests, this should be good to go from my side.
> 
> I see that a test is failing, is this unrelated noise?
> 
> ```
> Warning: Image did not match reference image /home/travis/build/juliohm/Variography.jl/test/data/Fitting.png. err: Images differ.  Difference: 0.022093271946049414  tolerance: 0.02
> ```

I will double check that as well. It may be related to changes in Plots.jl as usual.

File: src/empirical.jl (hunk 8 above)
Comment (visr) [reply src/empirical.jl]: It's a long list, I agree, though since it's meant as an internal function I think it's ok. We can shorten it only by adding more code duplication, which is also not great.

Do you want a leading underscore before internal function names by the way?

File: src/empirical.jl (hunk 9 above)
Comment (visr) [reply src/empirical.jl]: I did that here, where all the other keywords are set as well:

```julia
EmpiricalVariogram(sdata::AbstractData, var₁::Symbol, var₂::Symbol=var₁;
                   nlags=20, maxlag=nothing, distance=Euclidean(), nearestsearch=nothing) =
  EmpiricalVariogram(sdata, var₁, var₂, nlags, maxlag, distance, nearestsearch)
```

Not even sure if you can use keywords in a inner constructor actually.

**Comment by visr** [comment]
I don't have a strong preference on how to structure the code with the two pathways here. Purely looking at the amount of code duplication, I think the initial commit with two if-statements inside the for loop to handle the small differences was optimal. It maybe doesn't look efficient, but I'm quite sure the compiler optimizes them away (can benchmark of course).

**Comment by juliohm** [comment]
@visr could you share the real use cases you mentioned? I am doing some local benchmarks and the time difference is not that big with toy examples. I am also refactoring the code after a long time without touching it. :+1:

**Comment by visr** [comment]
Sure I can share an example. I'm not sure I can make it public at this point, so I'll send it to you in an email if that's ok.

If you want a toy example, you can also use my example with 10k random points above. The key to getting advantage from the spatial index is having a `max_lag` that includes a relatively small fraction of the points, for example for the random points between 0 and 1, I used 0.05 above, or 5% of the width of the data. If all points are always included in the range, the spatial index will in fact make it a bit slower, though I'm not sure how common that is.

**Comment by juliohm** [comment]
Thank you for the follow up. I will try to work on it a little bit further this morning. It seems like we need a different heuristic for the ball search mode. I am considering a volumetric one in which we check `volume(ball) <= volume(boundbox(sdata))` The right side of the inequality is already possible with the GeoStatsBase.jl API, but the left side needs an implementation. I will start adding a volume method for Euclidean balls, and we can add more in the future.

**Comment by visr** [comment]
Ok great, thanks. Such a volume based heuristic should work much better. I didn't pursue it because I was afraid it would be too expensive to calculate, but if it is not, we should go for it.

**Comment by juliohm** [comment]
We already have the bounding box calculation at the top of the empirical variogram calculation. It is costless depending on the spatial data type. For example, regular grids like images have a trivial bounding box. For arbitrary point sets, the algorithm is O(n) to find the bounding box. We should be fine.

**Comment by juliohm** [comment]
I've introduced a breaking change in GeoStatsBase.jl to construct neighborhood objects with an explicit dimension: https://github.com/juliohm/GeoStatsBase.jl/commit/1bd27c011aa2e1812fc1dfe82ef43875d90f451d

Now we need to specify the dimension of the space as a type parameter when constructing balls, for example `BallNeighborhood{3}(10.)` creates a 3D ball with radius 10 and Euclidean metric. The next step is to implement the volume of N-dimensional Euclidean balls using this formula: https://en.wikipedia.org/wiki/Volume_of_an_n-ball#The_volume

**Comment by juliohm** [comment]
Volume implemented: https://github.com/juliohm/GeoStatsBase.jl/commit/ab3e1d81ca0f51b3ddfd82cd6ffa3a90e0242306

Will get back to the variogram with ball search method soon.

**Comment by juliohm** [comment]
Thank you again @visr for this very nice contribution. I have pushed a bunch of commits to improve the code changes, and we now have an option `algo` that we can pass to the `EmpiricalVariogram` function to choose among `:full`, `:ball` or `:auto`. The `:auto` heuristic is implemented in terms of the `hmax` and `diagonal(boundbox(sdata))`. I was thinking about using the volume of the ball and the bounding box, but took the simpler heuristic for now. Please let me know if you feel that we should base the heuristic on volumes instead.

Next step to make this code faster is use multi-threading from Julia v1.3 :) Need to study the features of the language first for that, but I am positive we can gain some extra speed here.

Best,

**Comment by visr** [comment]
Great, thanks for finishing this! The `hmax` heuristic is indeed simpler and I think adequate in most cases. It's good to have a keyword to steer it.

One thing, looking at the docs:

https://github.com/juliohm/Variography.jl/blob/9d4bb027cfca467b002e3fbf408ac47939540b03/src/empirical.jl#L21-L27

It may not be clear to everyone that the choice here does not have any effect on the result, and can only affects the calculation speed. Reading that may put some people more at ease.

It also looks like the floating point requirement from here is now gone:

https://github.com/juliohm/Variography.jl/pull/8/commits/0d98370d2e4872d94b76479f466bbebb7fc0f208#diff-a0f56cf7c94c00247992a4ca3161738fR110-R112

Or does it actually work for integers?

**Comment by juliohm** [comment]
Very good point. I will write it explicitly in the docs that the result is identical. :100: 

I removed the check on the floating point bit as I think it may be fixed in NearestNeighbors.jl in the future? If users start reporting issues before that happens, we can add a warning or fix the selection of the algorithm to take this coordinate type into account. Restricting the algorithm on the coordinate type seems to harsh at first :)

**Comment by visr** [comment]
It will probably be fixed in the future, though note that https://github.com/KristofferC/NearestNeighbors.jl/issues/13 is from 2015.

I just tried with integer coordinates, and indeed you still get

```
MethodError: no method matching NearestNeighbors.KDTree(::Array{Int64,2}, ::Distances.Euclidean)
```

The reason I added this check was because this package's tests are using integer coordinates and started failing. Since `:auto` is the default, I worried it will break existing code.

Personally I don't care much about integer coordinates though, and I have no idea how common they are in practice.

**Comment by juliohm** [comment]
Thank you for checking it. I will add the check again before releasing a new version, you made a good point that user code may break otherwise.
